# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [14]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
load_dotenv()
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()
from newsapi import NewsApiClient

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\ljmac\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [15]:
# Read your api key environment variable
api_key = os.getenv("news_api")

In [16]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=api_key)

In [18]:
# Fetch the Bitcoin news articles
bitcoin_news = newsapi.get_everything(q="bitcoin", language="en", sort_by="relevancy")
bitcoin_news

{'status': 'ok',
 'totalResults': 7658,
 'articles': [{'source': {'id': None, 'name': 'Blogspot.com'},
   'author': 'noreply@blogger.com (Unknown)',
   'title': 'Bitcoin hits strongest level since May',
   'description': '<ul><li>Bitcoin, in terms of market value, rose 4.6% to $53,859.6.</li><li>It passed $50,000 mark for first time in four weeks on Tuesday.\xa0</li><li>Bitcoin fell below $50,000 in early September.</li></ul>Bitcoin hit its highest level since mid-May on Wednesda…',
   'url': 'https://techncruncher.blogspot.com/2021/10/bitcoin-hits-strongest-level-since-may.html',
   'urlToImage': 'https://blogger.googleusercontent.com/img/a/AVvXsEhiFZua2TKlWOHoZ5qipQ8zBVpSMbXNdxOFJmwNkddvji4GqJRO4SonCCmwV_IBrxtHCb2UlJ850RuZPY75s54QZkedzrYthaQqoo4tuOD3zUbMhTGxJ2T-rFtSu8I1vRJY9RtH8dWw7_5XtGfoeUvJuAWZU-U9XLsHKTKffBSWD6G9su34rjkO0sYD-w=w1200-h630-p-k-no-nu',
   'publishedAt': '2021-10-06T17:17:00Z',
   'content': '<ul><li>Bitcoin, in terms of market value, rose 4.6% to $53,859.6.</li><li>

In [19]:
# Fetch the Ethereum news articles
etherium_news = newsapi.get_everything(q="etherium", language="en", sort_by="relevancy")
etherium_news

{'status': 'ok',
 'totalResults': 26,
 'articles': [{'source': {'id': None, 'name': 'Yahoo Entertainment'},
   'author': 'Bryan Ke',
   'title': "'I have nothing left to live for': Evergrande meeting descends into chaos as investor pulls knife and threatens to kill herself",
   'description': 'A compilation video that shows desperate investors confronting Evergrande staff amid the company’s financial issues has gone viral on Chinese social media...',
   'url': 'https://news.yahoo.com/nothing-left-live-evergrande-meeting-174442622.html',
   'urlToImage': 'https://s.yimg.com/uu/api/res/1.2/KUw4qo9EIz7VbbKgYUv_fA--~B/aD00MjU7dz04MDA7YXBwaWQ9eXRhY2h5b24-/https://media.zenfs.com/en/nextshark_articles_509/483f64498cefe53217f4d52a4d99d33d',
   'publishedAt': '2021-10-01T17:44:42Z',
   'content': 'A compilation video that shows desperate investors confronting Evergrande staff amid the companys financial issues has gone viral on Chinese social media platform Weibo.\r\nThe confrontation: The 10-

In [20]:
# Create the Bitcoin sentiment scores DataFrame
bitcoin_sentiments = []

for article in bitcoin_news["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        bitcoin_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
    except AttributeError:
        pass

# Create DataFrame
bitcoin_df = pd.DataFrame(bitcoin_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
bitcoin_df = bitcoin_df[cols]

bitcoin_df.head()


date                                               text  compound  \
0  2021-10-06  <ul><li>Bitcoin, in terms of market value, ros...    0.3400   
1  2021-10-04  JPMorgan CEO Jamie Dimon is still not a Bitcoi...   -0.2411   
2  2021-10-06  How high are the chances of Bitcoin sustaining...    0.4939   
3  2021-10-10  Specifically, why did someone make a massive p...    0.5461   
4  2021-10-28  Theres a big new presence slurping up power fr...    0.3612   

   positive  negative  neutral  
0     0.076     0.000    0.924  
1     0.000     0.116    0.884  
2     0.219     0.000    0.781  
3     0.121     0.000    0.879  
4     0.096     0.000    0.904

In [21]:
# Create the Ethereum sentiment scores DataFrame
etherium_sentiments = []

for article in etherium_news["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        etherium_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
    except AttributeError:
        pass

# Create DataFrame
etherium_df = pd.DataFrame(etherium_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
etherium_df = etherium_df[cols]

etherium_df.head()

date                                               text  compound  \
0  2021-10-01  A compilation video that shows desperate inves...   -0.6369   
1  2021-10-20  AppleInsider is supported by its audience and ...    0.4588   
2  2021-10-01  Lindsay Lohans had a busy year for someone who...    0.2960   
3  2021-10-01  Mike Persichetti, the sales manager at Cocoas ...    0.8591   
4  2021-10-07  It’s been a good week for cryptocurrency trade...    0.5719   

   positive  negative  neutral  
0     0.000     0.193    0.807  
1     0.121     0.000    0.879  
2     0.058     0.000    0.942  
3     0.226     0.028    0.746  
4     0.139     0.000    0.861

In [22]:
# Describe the Bitcoin Sentiment
bitcoin_df.describe()

compound   positive   negative   neutral
count  20.000000  20.000000  20.000000  20.00000
mean    0.210935   0.066850   0.008850   0.92435
std     0.296789   0.076818   0.028664   0.07608
min    -0.241100   0.000000   0.000000   0.78100
25%     0.000000   0.000000   0.000000   0.88275
50%     0.012900   0.041000   0.000000   0.91400
75%     0.467575   0.107500   0.000000   1.00000
max     0.755800   0.219000   0.116000   1.00000

In [23]:
# Describe the Ethereum Sentiment
etherium_df.describe()

compound   positive   negative    neutral
count  20.000000  20.000000  20.000000  20.000000
mean    0.323700   0.118250   0.044750   0.837000
std     0.440385   0.067113   0.066367   0.078235
min    -0.750600   0.000000   0.000000   0.712000
25%     0.158175   0.076500   0.000000   0.786500
50%     0.449600   0.120500   0.000000   0.847000
75%     0.589275   0.139000   0.072500   0.879250
max     0.888500   0.258000   0.220000   1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: The highest mean positive score was Etherium with a total of 0.11825

Q: Which coin had the highest compound score?

A: The highest (max) compound score was Etherium again with a total of 0.8885

Q. Which coin had the highest positive score?

A: Etherium had the highest (max) positive score coming in at 0.258

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [24]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [37]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
sw = set(stopwords.words('english'))


# Expand the default stopwords list if necessary
#sw_addon = {'said', 'mln', 'kilolitres','kl'}
#second_result = [word.lower() for word in words if word.lower() not in sw.union(sw_addon)]



A bitcoin bull is on pullback watch.
Hedge fund manager Mark Yusko believes investors will take profits due to the cryptocurrency's sharp rally over the last few weeks.
"There are a lot of people t… [+2258 chars]


In [44]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    
    # Remove the punctuation from text
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)

    # Create a tokenized list of the words
    words = word_tokenize(re_clean)
    
    # Lemmatize words into root words
    lem = [lemmatizer.lemmatize(word) for word in words]
   
    # Convert the words to lowercase
    tokens = [word.lower() for word in lem if word.lower() not in sw]
    

    # Remove the stop words
    
    
    return tokens
    

In [13]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
all_btc_text = btc.text.str.cat()
all_btc_text

In [28]:
# Run the NER processor on all of the text
doc = nlp(all_btc_text)

# Add a title to the document
doc.user_data["title"] = "Bitcoin NER"

In [29]:
# Render the visualization
displacy.render(doc, style='ent', jupyter=True)

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
all_eth_text = eth.text.str.cat()
all_eth_text

In [32]:
# Run the NER processor on all of the text
doc = nlp(all_eth_text)

# Add a title to the document
doc.user_data["title"] = "Etherium NER"

In [33]:
# Render the visualization
displacy.render(doc, style='ent', jupyter=True)

In [34]:
# List all Entities
# YOUR CODE HERE!

---